Iss388 cvt cosmics

zieglerv opened the pull request.

Commit:
Bug fix in setting the measurements for cosmics.  Added missing hemisphere for BST clusters.

Commit:
fix in propagation to surface for shifted cylinders.

Commit:
Fix measurements sorting by Y intersection of the cosmic ray with the surface.

Commit:
removed unused import

Commit:
Initialized chi^2

Commit:
Added y intersect for rays.  This is used to sort the surfaces from top to bottom.
In the case of a shallow angle ray there can be 2 intersection with a passive surface in the same hemisphere, hence the need to sort by y.

Commit:
For cosmic fitting, save the output for the best chi^2; do not kill the track if the next iteration chi^2 is worse.

Commit:
Use global fit y-intersection with the surface to assign fitted track intersection point when there are 2 intersection points of the ray with the cylinder.

Commit:
Assign correct hemisphere for cosmics.
Use cluster information to determine y-intersection with the surface and hemisphere for active surfaces.
Compute the  y-intersection with the surface to be used for sorting in the kalman fitter.

Commit:
Cosmic debug statement

raffaelladevita commented on common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/Surface.java:
```diff
@@ -42,6 +42,7 @@ public class Surface implements Comparable<Surface> {
     public double swimAccuracy;
     public boolean passive = false;
     public double hemisphere = 1;
+    public double rayYinterc = 9999;
```
let's make it inf or nan

raffaelladevita replied:
Also, to be more general, it could make sense to add a Point3D to store the full information of the trajectory point instead of only y

raffaelladevita commented on common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/helical/StateVecs.java:
```diff
@@ -53,22 +53,31 @@ public boolean getStateVecPosAtMeasSite(StateVec sv, AMeasVecs.MeasVec mv, Swim
-                sv.z = inters.z();
-                sv.path = inters.distance(st);
-            }
+//                mv.surface.toLocal().apply(st);
```
let's remove the commented-out lines...they will still be in the history

raffaelladevita commented on common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/helical/StateVecs.java:
```diff
@@ -53,22 +53,31 @@ public boolean getStateVecPosAtMeasSite(StateVec sv, AMeasVecs.MeasVec mv, Swim
+                int ints = mv.surface.cylinder.intersection(toPln, inters);
+                if(ints<1) return false;
+               
+                    sv.x = inters.get(0).x()  ;
```
is the first intersection always the right one?

raffaelladevita commented on common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/straight/KFitter.java:
```diff
@@ -54,22 +54,24 @@ public void runFitter(AStateVecs sv, AMeasVecs mv) {
                 this.setTrajectory(sv, mv);
-                setFitFailed = false;
+                //System.out.println(finalStateVec.toString());
+                //setFitFailed = false;
```
I assume setFitFailed is not really used elsewhere but it may be better to still set it to false
(and remove the commented text)

raffaelladevita commented on common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/straight/StateVecs.java:
```diff
@@ -43,22 +44,41 @@ else if(mv.hemisphere!=0) {
+                } else {
+                    int index =0;
+                    if(ints>1) {
+                        double dh0 = Math.abs(inters.get(0).y()-mv.surface.rayYinterc);
```
Shouldn't the right intersection be the one with the largest y among the ones with y<rayInterc?

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/measurement/Measurements.java:
```diff
@@ -229,18 +232,30 @@ public List<Surface> getMeasurements(StraightTrack cosmic) {
                 }
-                surfaces.add(surf);
+                
+                sMap.put(surf.getIndex(), surf);
```
is this changing something?

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/services/CVTEngine.java:
these changes are not really necessary

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/measurement/Measurements.java:
```diff
@@ -425,25 +443,149 @@ else if(type==DetectorType.BMT) {
+                return hemisphereSign == (int) Math.signum(intersections.get(0).y());
+            case 2:
+                // Check if either of the two intersections match the hemisphere
+                return hemisphereSign == (int) Math.signum(intersections.get(0).y()) ||
```
for small y, isn't possible the sign doesn't match because of resolution?

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/measurement/Measurements.java:
```diff
@@ -425,25 +443,149 @@ else if(type==DetectorType.BMT) {
-        if(surface.cylinder==null)
-            return 0;
-        List<Point3D> trajs = new ArrayList<>();
+    private double handleCylinderIntersection(int h, Cylindrical3D cylinder, Line3D line) {
```
h could be renamed as hemisphere for clarity

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/measurement/Measurements.java:
```diff
@@ -425,25 +443,149 @@ else if(type==DetectorType.BMT) {
+        double Y = this.getY(cluster, ray, surface);
+        if(Y == Double.POSITIVE_INFINITY) return 0;
+        surface.rayYinterc=Y;
+        return (int) Math.signum(Y);
```
if misalignments are large in y, the sign of the y coordinate of the intersection point doesn't really correspond to the hemisphere

raffaelladevita reviewed (changes requested):
it looks to me as if the main fix is in the calculation of the trajectory point that doesn't assume anymore the cylinders representing the detector surfaces are centered at 0,0, but I'm unsure about some of the changes to the measurement class

Commit:
Change intersection of ray with surface from y-coordinate to Point3D.
Change isCrossed method to use the value of the hemisphere which is calculated 
as the sign of the y value of the intersection of the ray with the surface 
if it crosses it or 0 if it does not.

Commit:
Remove commented out old code.
Fix intersection of ray with cylinder for tracks from the target.
Intersection initialized as positive infinity.

raffaelladevita commented on reconstruction/cvt/src/main/java/org/jlab/rec/cvt/measurement/Measurements.java:
```diff
@@ -283,11 +297,14 @@ private void addClusters(DetectorType type, List<Surface> clusters) {
         
     private List<Surface> getClusterSurfaces(DetectorType type, List<Cluster> clusters, Ray ray) {
-        
-        List<Surface> surfaces = this.getClusterSurfaces(type, clusters, ray);        
```
it seems the initial source of the problem is this line, a call to the method from the same method. This was introduced with #351.

Fixing this and reverting to assigning the hemisphere based on the cluster.y instead of with the ray intersection seems to solve the problem.

The ray intersection with the surface doesn't seem to work consistently because when two intersection points are found (for cylindrical surfaces) the first one in the List is not necessarily the closest to the ray origin. This is the case also with intersectionRay
